internal/gatewayapi/route.go

```diff
@@ -615,76 +621,99 @@ func (t *Translator) processGRPCRouteParentRefs(grpcRoute *GRPCRouteContext, res
 		// a unique Xds IR HTTPRoute per match.
 		ruleRoutes, err := t.processGRPCRouteRule(grpcRoute, ruleIdx, httpFiltersContext, rule)
 		if err != nil {
-			return nil, err
+			errs.Add(status.NewRouteStatusError(
+				fmt.Errorf("failed to process route rule %d: %w", ruleIdx, err),
+				status.ConvertToAcceptedReason(err.Reason()),
+			).WithType(gwapiv1.RouteConditionAccepted))
+			continue
 		}
 
 		destName := irRouteDestinationName(grpcRoute, ruleIdx)
+		allDs := []*ir.DestinationSetting{}
+		failedProcessDestination := false
+
 		for i, backendRef := range rule.BackendRefs {
 			settingName := irDestinationSettingName(destName, i)
 			ds, err := t.processDestination(settingName, backendRef, parentRef, grpcRoute, resources)
-			if ds == nil {
+			if err != nil {
+				errs.Add(status.NewRouteStatusError(
+					fmt.Errorf("failed to process route rule %d backendRef %d: %w", ruleIdx, i, err),
+					err.Reason(),
+				))
+				failedProcessDestination = true
 				continue
 			}
 
-			for _, route := range ruleRoutes {
-				// If the route already has a direct response or redirect configured, then it was from a filter so skip
-				// processing any destinations for this route.
-				if route.DirectResponse != nil || route.Redirect != nil {
-					continue
-				}
-
-				//  disable associated routes to a backend ref in case some of its config was invalid
-				if err != nil {
-					route.DirectResponse = &ir.CustomResponse{
-						StatusCode: ptr.To(uint32(500)),
-					}
-				}
-
-				if route.Destination == nil {
-					route.Destination = &ir.RouteDestination{
-						Name: destName,
-					}
-				}
-				route.Destination.Settings = append(route.Destination.Settings, ds)
+			if ds == nil {
+				continue
 			}
+			allDs = append(allDs, ds)
 		}
 
-		// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
-		for _, ruleRoute := range ruleRoutes {
-			noValidBackends := ruleRoute.Destination == nil || ruleRoute.Destination.ToBackendWeights().Valid == 0
-			if noValidBackends && ruleRoute.Redirect == nil {
-				ruleRoute.DirectResponse = &ir.CustomResponse{
+		// process each ir route
+		for _, irRoute := range ruleRoutes {
+			irRoute.IsHTTP2 = true
+			destination := &ir.RouteDestination{
+				Settings: allDs,
+			}
+
+			switch {
+			// If the route already has a direct response or redirect configured, then it was from a filter so skip
+			// processing any destinations for this route.
+			case irRoute.DirectResponse != nil || irRoute.Redirect != nil:
+			// return 500 if any destination setting is invalid
+			// the error is already added to the error list when processing the destination
+			case failedProcessDestination:
+				irRoute.DirectResponse = &ir.CustomResponse{
+					StatusCode: ptr.To(uint32(500)),
+				}
+			// return 500 if the weight of all the valid destination settings(endpoints list is not empty) is 0
+			case destination.ToBackendWeights().Valid == 0:
```

Comment:
what about the case when backend has no endpoints, do we end up here ?

Reply:
The pervious behavior for backend without endpoints is 500 direct response, and no error status for Accepted and Resolved conditions.  This PR doesn't change it.